packages/common-contracts/test/erc20.spec.ts

```diff
@@ -0,0 +1,507 @@
+  })
+
+  let Factory__TestERC20: ContractFactory
+  before(async () => {
+    Factory__TestERC20 = await l2ethers.getContractFactory('TestERC20')
+  })
+
+  let TestERC20: Contract
+  beforeEach(async () => {
+    TestERC20 = await Factory__TestERC20.connect(initialHolder).deploy(
+      initialSupply,
+      tokenName,
+      tokenSymbol
+    )
+    await sleep(100)
+    await TestERC20.deployTransaction.wait()
+  })
+
+  it('should have the right initial supply', async () => {
+    const totalSupply = await TestERC20.totalSupply()
+    expect(totalSupply.toNumber()).to.equal(initialSupply)
+  })
+
+  it('should have the right token name', async () => {
+    const name = await TestERC20.name()
+    expect(name).to.equal(tokenName)
+  })
+
+  it('should have the right token symbol', async () => {
+    const symbol = await TestERC20.symbol()
+    expect(symbol).to.equal(tokenSymbol)
+  })
+
+  it('should have the default 18 decimal places', async () => {
+    const decimals = await TestERC20.decimals()
+    expect(decimals).to.equal(18)
+  })
+
+  it('should assign initial balance', async () => {
+    const balance = await TestERC20.balanceOf(initialHolder.address)
+    expect(balance.toNumber()).to.equal(initialSupply)
+  })
+
+  describe('decrease allowance', () => {
+    describe('when the spender is not the zero address', () => {
+      let spender: string
+      beforeEach(async () => {
+        spender = recipient.address
+      })
+
+      const shouldDecreaseApproval = (amount: number) => {
+        describe('when there was no approved amount before', () => {
+          it.skip('reverts', async () => {
+            await expect(
+              TestERC20.decreaseAllowance(spender, amount)
+            ).to.be.revertedWith('ERC20: decreased allowance below zero')
+          })
+        })
+
+        describe('when the spender had an approved amount', () => {
+          const approvedAmount = amount
+
+          beforeEach(async () => {
+            const tx = await TestERC20.approve(spender, approvedAmount)
+            await sleep(100)
+            await tx.wait()
+          })
+
+          it('emits an approval event', async () => {
+            await expect(TestERC20.decreaseAllowance(spender, approvedAmount))
+              .to.emit(TestERC20, 'Approval')
+              .withArgs(initialHolder.address, spender, 0)
+          })
+
+          it('decreases the spender allowance subtracting the requested amount', async () => {
+            await TestERC20.decreaseAllowance(spender, approvedAmount - 1)
+
+            expect(
+              (
+                await TestERC20.allowance(initialHolder.address, spender)
+              ).toNumber()
+            ).to.equal(1)
+          })
+
+          it('sets the allowance to zero when all allowance is removed', async () => {
+            await TestERC20.decreaseAllowance(spender, approvedAmount)
+
+            expect(
+              (
```

Comment:
Could you make this less nested? It makes it hard to read when its nested and indented a bunch

Reply:
```js
const allowance = await TestERC20.allowance(initialHolder.address, spender)
expect(allowance.toNumber()).to.equal(0)
```

Reply:
Its also less lines